File: src/pocketcloud/cloud/template/TemplateHelper.php
<?php

namespace pocketcloud\cloud\template;

use pocketcloud\cloud\util\Utils;

final class TemplateHelper {

    public const array KEYS = ["name", "lobby", "maintenance", "static", "maxPlayerCount", "minServerCount", "maxServerCount", "startNewPercentage", "autoStart", "templateType"];
    public const array EDITABLE_KEYS = ["lobby", "maintenance", "static", "maxPlayerCount", "minServerCount", "maxServerCount", "startNewPercentage", "autoStart"];

    public const array NECESSARY_KEYS = ["name", "lobby", "templateType"];

    public const array UNNECESSARY_KEYS = ["maintenance", "static", "maxPlayerCount", "minServerCount", "maxServerCount", "startNewPercentage", "autoStart"];
    public const array DEFAULT_VALUES = ["maintenance" => true, "static" => false, "mayPlayerCount" => 20, "minServerCount" => 0, "maxServerCount" => 2, "startNewPercentage" => 100, "autoStart" => true];
    private const array CONVERSION = [
        "maxplayercount" => "maxPlayerCount",
        "minservercount" => "minServerCount",
        "maxservercount" => "maxServerCount",
        "startnewpercentage" => "startNewPercentage",
        "autostart" => "autoStart"
    ];

    public static function addUnnecessaryKeys(array $data): void {
        foreach (array_filter(self::UNNECESSARY_KEYS, fn(string $key) => !isset($data[$key])) as $key) $data[$key] = self::DEFAULT_VALUES[$key];
    }

    public static function sumSettingsToInstance(array $data): ?TemplateSettings {
        if (Utils::containKeys($data, ...self::EDITABLE_KEYS)) {
            $onlySettings = [];
            foreach (self::EDITABLE_KEYS as $key) $onlySettings[$key] = $data[$key];
            return TemplateSettings::fromArray($onlySettings);
        }
        return null;
    }

    public static function isValidEditValue(string $value, string $key, ?string &$expected = null, mixed &$realValue = null): bool {
        if ($key == "lobby" || $key == "maintenance" || $key == "autoStart" || $key == "static") {
            $expected = "true | false";
            if ($value == "true" || $value == "false") {
                $realValue = $value == "true";
                return true;
            }
        } else if ($key == "maxPlayerCount" || $key == "minServerCount" || $key == "maxServerCount" || $key == "startNewPercentage") {
            $expected = "number";
            if (is_numeric($value)) {
                $realValue = max($key == "startNewPercentage" ? floatval($value) : intval($value), 0);
                return true;
            }
        }
        return false;
    }

    public static function isValidEditKey(string $key): bool {
        return in_array($key, self::EDITABLE_KEYS);
    }

    public static function convert(string $key): string {
        return self::CONVERSION[strtolower($key)] ?? $key;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/template/util/TemplateHelper.java
package de.pocketcloud.cloud.template.util;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

public final class TemplateHelper {

    public static final List<String> KEYS = List.of(
            "name", "lobby", "maintenance", "staticServers", "alwaysCopyToStaticServers", "saveOnShutdown",
            "maxPlayerCount", "minServerCount", "maxServerCount", "startNewPercentage",
            "autoStart", "templateType", "serverSoftware", "maxMemory"
    );

    public static final List<String> EDITABLE_KEYS = List.of(
            "lobby", "maintenance", "staticServers", "alwaysCopyToStaticServers", "saveOnShutdown",
            "maxPlayerCount", "minServerCount", "maxServerCount",
            "startNewPercentage", "autoStart", "maxMemory"
    );

    public static final Map<String, Class<?>> KEY_TYPES = new HashMap<>();

    private static final Map<String, String> CONVERSION = Map.of(
            "staticservers", "staticServers",
            "alwayscopytostaticservers", "alwaysCopyToStaticServers",
            "maxplayercount", "maxPlayerCount",
            "minservercount", "minServerCount",
            "maxservercount", "maxServerCount",
            "startnewpercentage", "startNewPercentage",
            "autostart", "autoStart",
            "templatetype", "templateType",
            "serversoftware", "serverSoftware"
    );

    static {
        KEY_TYPES.put("lobby", Boolean.class);
        KEY_TYPES.put("maintenance", Boolean.class);
        KEY_TYPES.put("staticServers", Boolean.class);
        KEY_TYPES.put("alwaysCopyToStaticServers", Boolean.class);
        KEY_TYPES.put("saveOnShutdown", Boolean.class);
        KEY_TYPES.put("maxPlayerCount", Integer.class);
        KEY_TYPES.put("minServerCount", Integer.class);
        KEY_TYPES.put("maxServerCount", Integer.class);
        KEY_TYPES.put("startNewPercentage", Double.class);
        KEY_TYPES.put("autoStart", Boolean.class);
        KEY_TYPES.put("templateType", String.class);
        KEY_TYPES.put("serverSoftware", String.class);
        KEY_TYPES.put("maxMemory", Integer.class);
    }

    private static final Set<String> EDITABLE_KEYS_SET = Set.copyOf(EDITABLE_KEYS);

    public static String convert(String key) {
        return CONVERSION.getOrDefault(key.toLowerCase(), key);
    }

    public static boolean checkKey(String key) {
        return EDITABLE_KEYS_SET.contains(key);
    }

    public static Class<?> getKeyType(String key) {
        return KEY_TYPES.get(key);
    }
}